docs/operations.md

```diff
@@ -28,6 +28,138 @@ Clients may inspect this specification at run-time so to obey the various limits
+
+The event stream looks as follows:
+
+```
+data: [...]
+id: 1
+
+data: [...]
+id: 2
+
+[...]
+```
+
+with a single event being structured as follows:
+
+```
+data: { "action": "<ACTION>", "job": <JOB> }
+id: <EVENT_ID>
+
+```
+
+**Note**: Each event is terminated by a pair of newline characters `\n\n` (as required by the SSE spec).
+
+The semantics of the individual fields is:
+
+- `<ACTION>` specifies the type of event that occurred. The valid actions are:
+  - `CREATE`: a new job has been created
+  - `DELETE`: an existing job has been deleted
+  - `ADD_TAGS`: tags were added to a job
+  - `DELETE_TAGS`: tags were removed from a job
+  - `UPDATE_STATUS`: job status has been updated
+  - `UPDATE_DEFINITION`: job definition has been updated
+- `<JOB>` is the JSON representation of the job associated with this event
+- `<EVENT_ID>`: a (unique) numeric identifier for the event, starting at 1 and incremented by 1 for each subsequent event; can be used to detect message loss on the client side.
+
+**Example:**
+
+```
+data: {"action":"UPDATE_STATUS","job":{"clientId":"Dana","definition":{"title":"Expose Job API"},"history":null,"id":"d305e539-1d41-4c95-b19a-2a7055c469d0","mtime":"2023-08-29T13:00:57.632+02:00","status":{"definitionHash":"f8b3d6ea0150a98ec5f8dc224659211a37cb4737777388f77a3e323a865e7674","state":"PROGRESS"},"stime":"2023-08-22T15:30:02.323+02:00","tags":[],"workflow":{"groups":null,"name":"wfx.workflow.kanban","states":null,"transitions":null}}}
+id: 1
+
+```
+
+#### Filter Parameters
+
+Job events can be filtered using any combination of the following parameters:
+
+- Job IDs
+- Client IDs
+- Workflow Names
+
+This enables more precise control over the events which are dispatched.
+
+#### Examples
+
+`wfxctl` offers a reference client implementation. The following command subscribes to **all** job events:
+
+```bash
+wfxctl job events
+```
+
+This may result in a large number of events, though. For a more targeted approach, filter parameters may be used.
+Assuming the job IDs are known (either because the jobs have been created already or the IDs are received via another
+subscription channel), the following will subscribe to events matching either of the two specified job IDs:
+
+```bash
+wfxctl job events --ids d305e539-1d41-4c95-b19a-2a7055c469d0,e692ad92-45e6-4164-b3fd-8c6aa884011c
+```
+
+See `wfxctl job events --help` for other filter parameters, e.g. workflow names.
+
+#### Considerations and Limitations
+
+1. **Asynchronous Job Status Updates**: Job status updates are sent asynchronously to avoid the risk of a subscriber
+   causing delays. As a result, messages could arrive in a not necessarily linear order, deviating from the client's
+   expectation of a naturally ordered flow of events. While this is generally not a concern, it could become an issue in
+   high-concurrency situations. For example, when multiple clients try to modify the same job or when a single client
+   issues a rapid sequence of status updates.
```

Comment:
... Then, the event's `id:` field can be considered by the client to establish it's linear order view of events.

Reply:
Unfortunately, that's not the case. We have a single writer thread (goroutine) that generates the IDs for each event before submitting them to the client. As a result, the IDs will always appear in natural order. However, there can be multiple concurrent threads that update a job's status, for example. Each of these update threads enqueues a message for the writer thread. In reality, we don't know the precise order in which these updates are applied to the database. Each update operates within a transaction, and if the commit succeeds, we consider it acceptable. Even if we generate the ID immediately after a successful commit, there's no guarantee that it reflects the actual sequence of events, as two threads may race to obtain the ID.
As for the primary purpose of the ID, my understanding is that it is used to resume an interrupted connection between the client and server. The server would then resend messages starting from a specified ID. However, we don't currently support this feature since we don't retain any information about previous clients. Nonetheless, I believe the ID is useful for quality of service (QoS) and for detecting network delays or issues as you outlined in one of the previous comments :)